demo-circuit/src/naive_threshold_sig_w_key_rotation/tests.rs

```diff
@@ -0,0 +1,1180 @@
+                    num_sig_changed, valid_sigs
+                );
+            }
+            KeyToChange::MasterKey => {
+                circuit.validator_keys_signatures.updated_master_keys[num_mst_changed] = rng.gen();
+                println!(
+                    "num_mst_changed: {:?}, valid_sigs: {:?}",
+                    num_mst_changed, valid_sigs
+                );
+            }
+            KeyToChange::SigningAndMasterKeys => {
+                circuit.validator_keys_signatures.updated_signing_keys[num_sig_changed] = rng.gen();
+                circuit.validator_keys_signatures.updated_master_keys[num_mst_changed] = rng.gen();
+                println!(
+                    "num_sig_changed: {:?}, num_mst_changed: {:?}, valid_sigs: {:?}",
+                    num_sig_changed, num_mst_changed, valid_sigs
+                );
+            }
+        }
+
+        let failing_constraint = test_naive_threshold_circuit(
+            circuit.clone(),
+            debug_only,
+            max_pks,
+            num_custom_fields,
+            None,
+        );
+
+        println!("Failing constraint: {:?}", failing_constraint);
+        assert!(
+            failing_constraint
+                .clone()
+                .unwrap()
+                .contains("check updated signing key should be signed old signing key ")
+                || failing_constraint
+                    .clone()
+                    .unwrap()
+                    .contains("check updated master key should be signed old signing key ")
+        );
+    }
+}
+
+#[serial]
+#[test]
+fn test_naive_threshold_sig_w_key_rotation_circuit_wrong_scb_validators_keys_root_prev_cert_false()
+{
+    let debug_only = true;
+
+    let (
+        max_pks,
+        num_custom_fields,
+        valid_sigs,
+        threshold,
+        is_first_certificate,
+        expected_threshold_results,
+        change_signing_keys,
+        change_master_keys,
+    ) = generate_circuit_test_data(TestDataType::AllRandomFirstCert);
+    assert!(is_first_certificate == true);
+    assert!(expected_threshold_results == true);
+
+    println!("Checking wrong prev_scb_validators_keys_root when is_first_certificate is true");
+    let mut circuit = get_test_circuit_instance(
+        max_pks,
+        num_custom_fields,
+        valid_sigs,
+        threshold,
+        is_first_certificate,
+        change_signing_keys,
+        change_master_keys,
+    );
+
+    let rng = &mut thread_rng();
+
+    //scb_validators_keys_root is in first prev custom fields
+    circuit.withdrawal_certificate.custom_fields[0] = rng.gen();
+
+    let failing_constraint =
+        test_naive_threshold_circuit(circuit, debug_only, max_pks, num_custom_fields, None);
+
+    println!("Failing constraint: {:?}", failing_constraint);
+    assert!(failing_constraint
+        .clone()
+        .unwrap()
+        .contains("valid_signatures == quality/conditional_equals"));
```

Comment:
Of course changing custom_fields[0] changes certificate hash thus it invalidates the signatures too, so we are not testing the constraint that is supposed to fail. Perhaps we can move this check to end of the circuit too ?

Reply:
The problem is that we are not asserting that the correct constraint fails considering what we want to test, as the debug_circuit function only returns the first unsatisfied constraint and, since the check on the quality comes before the check on the validator keys root, the debug_circuit function will return the former